packages/frontend/core/src/components/attachment-viewer/worker/worker.ts

```diff
@@ -0,0 +1,223 @@
+  const oldEndIndex = ranges[`${kind}:endIndex`];
+  let i = 0;
+  let l = 0;
+
+  if (oldEndIndex < startIndex || oldStartIndex > endIndex) {
+    i = oldStartIndex;
+    l = oldEndIndex;
+  } else {
+    const oldMid = Math.ceil((oldStartIndex + oldEndIndex) / 2);
+    const mid = Math.ceil((startIndex + endIndex) / 2);
+    const diff = Math.abs(mid - oldMid);
+
+    if (mid > oldMid) {
+      i = oldStartIndex;
+      l = i + diff;
+    } else if (mid < oldMid) {
+      i = endIndex;
+      l = i + diff;
+    }
+  }
+
+  for (; i < l + 1; i++) {
+    cached.delete(`${kind}:${i}`);
+  }
+
+  ranges[`${kind}:startIndex`] = startIndex;
+  ranges[`${kind}:endIndex`] = endIndex;
+}
+
+async function start() {
+  inited = true;
+
+  logger.debug('pdf worker pending');
+  self.postMessage({ type: MessageOp.Init });
+
+  const pdfium = await createPDFium();
+  viewer = new Viewer(new Runtime(pdfium));
+
+  self.postMessage({ type: MessageOp.Inited });
+  logger.debug('pdf worker ready');
+}
+
+async function process({ data }: MessageEvent<MessageData>) {
+  if (!inited) {
+    await start();
+  }
+
+  if (!viewer) return;
+
+  const { type } = data;
+
+  switch (type) {
+    case MessageOp.Open: {
+      const buffer = data[type];
+      if (!buffer) return;
+
+      doc = viewer.open(new Uint8Array(buffer));
+
+      if (!doc) return;
+
+      const page = doc.page(0);
+
+      if (!page) return;
+
+      Object.assign(docInfo, {
+        total: doc.pageCount(),
+        height: Math.ceil(page.height()),
+        width: Math.ceil(page.width()),
+      });
+      page.close();
+      post(MessageOp.Opened, docInfo);
+
+      break;
+    }
+
+    case MessageOp.Render: {
+      if (!doc) return;
+
+      const {
+        kind,
+        scale,
+        range: { startIndex, endIndex },
+      } = data[type];
+
+      if (startIndex > endIndex || startIndex < 0) return;
+
+      const { total } = docInfo;
+      const queue: (() => Promise<void | [void, void]>)[] = [];
+
+      if (startIndex === 0) {
+        for (let n = startIndex; n <= endIndex; n++) {
+          const b = createJob(n, kind, scale);
+          queue.push(b);
+        }
+      } else if (endIndex + 1 === total) {
+        for (let n = endIndex; n >= startIndex; n--) {
+          const a = createJob(n, kind, scale);
+          queue.push(a);
+        }
+      } else {
+        const mid = Math.floor((startIndex + endIndex) / 2);
+        const m = createJob(mid, kind, scale);
+        queue.push(m);
+
+        let n = 1;
+        const s = Math.max(endIndex - mid, mid - startIndex);
+        for (; n <= s; n++) {
+          const j = Math.min(mid + n, endIndex);
+          const i = Math.max(mid - (j - mid), 0);
+          const a = createJob(j, kind, scale);
+          const b = createJob(i, kind, scale);
+          const ab = () => Promise.all([a(), b()]);
+          queue.push(ab);
+        }
+      }
+
+      queueMicrotask(() => {
```

Comment:
queue the job one by one, and notify parent process by `postMessage` when image rendered